File: backbone-go/domain/exceptions/domain_exceptions.go
// Package exceptions provides domain layer exceptions with 8-digit error codes (11xxxxxx)
package exceptions

import (
	"encoding/json"
	"fmt"
	"time"
)

// ErrorSeverity represents the severity level of an error
type ErrorSeverity string

const (
	SeverityCritical ErrorSeverity = "critical"
	SeverityHigh     ErrorSeverity = "high"
	SeverityMedium   ErrorSeverity = "medium"
	SeverityLow      ErrorSeverity = "low"
)

// BaseKernelException is the base exception for all framework exceptions
type BaseKernelException struct {
	Code        int                    `json:"code"`
	Message     string                 `json:"message"`
	HTTPCode    int                    `json:"http_code"`
	Details     map[string]interface{} `json:"details,omitempty"`
	Layer       string                 `json:"layer"`
	Severity    ErrorSeverity          `json:"severity"`
	Timestamp   time.Time              `json:"timestamp"`
	Recoverable bool                   `json:"recoverable"`
	RetryAfter  *int                   `json:"retry_after,omitempty"`
	Context     map[string]interface{} `json:"context,omitempty"`
	OriginalErr error                  `json:"-"`
}

// Error implements the error interface
func (e *BaseKernelException) Error() string {
	return fmt.Sprintf("[%s-%d] %s", e.Layer, e.Code, e.Message)
}

// ToJSON converts exception to JSON string
func (e *BaseKernelException) ToJSON() string {
	bytes, _ := json.MarshalIndent(e, "", "  ")
	return string(bytes)
}

// ToDict converts exception to map
func (e *BaseKernelException) ToDict() map[string]interface{} {
	result := map[string]interface{}{
		"error_code":    e.Code,
		"error_message": e.Message,
		"http_code":     e.HTTPCode,
		"layer":         e.Layer,
		"severity":      e.Severity,
		"timestamp":     e.Timestamp.Format(time.RFC3339),
		"recoverable":   e.Recoverable,
	}

	if e.Details != nil && len(e.Details) > 0 {
		result["error_details"] = e.Details
	}

	if e.RetryAfter != nil {
		result["retry_after"] = *e.RetryAfter
	}

	if e.Context != nil && len(e.Context) > 0 {
		result["context"] = e.Context
	}

	return result
}

// DomainException represents domain layer exceptions (11xxxxxx)
type DomainException struct {
	*BaseKernelException
}

// NewDomainException creates a new domain exception
func NewDomainException(code int, message string, details map[string]interface{}) *DomainException {
	if code < 11000000 || code > 11999999 {
		panic(fmt.Sprintf("Domain exception code must be in range 11000000-11999999, got: %d", code))
	}

	return &DomainException{
		BaseKernelException: &BaseKernelException{
			Code:        code,
			Message:     message,
			HTTPCode:    400,
			Details:     details,
			Layer:       "domain",
			Severity:    SeverityHigh,
			Timestamp:   time.Now().UTC(),
			Recoverable: false,
			Context:     make(map[string]interface{}),
		},
	}
}

// BusinessRuleViolationException for business rule violations
type BusinessRuleViolationException struct {
	*DomainException
	RuleName string
}

// NewBusinessRuleViolationException creates a business rule violation exception
func NewBusinessRuleViolationException(message, ruleName string) *BusinessRuleViolationException {
	details := map[string]interface{}{
		"rule_violated": ruleName,
	}

	exception := NewDomainException(11001001, message, details)

	return &BusinessRuleViolationException{
		DomainException: exception,
		RuleName:        ruleName,
	}
}

// InvalidEntityStateException for invalid entity states
type InvalidEntityStateException struct {
	*DomainException
	EntityType    string
	EntityID      string
	CurrentState  string
	RequiredState string
}

// NewInvalidEntityStateException creates an invalid entity state exception
func NewInvalidEntityStateException(
	message, entityType, entityID, currentState, requiredState string,
) *InvalidEntityStateException {
	details := map[string]interface{}{
		"entity_type":    entityType,
		"entity_id":      entityID,
		"current_state":  currentState,
		"required_state": requiredState,
	}

	exception := NewDomainException(11002001, message, details)
	exception.HTTPCode = 412 // Precondition Failed

	return &InvalidEntityStateException{
		DomainException: exception,
		EntityType:      entityType,
		EntityID:        entityID,
		CurrentState:    currentState,
		RequiredState:   requiredState,
	}
}

// InvalidValueObjectException for invalid value objects
type InvalidValueObjectException struct {
	*DomainException
	ValueObjectType string
	InvalidValue    interface{}
}

// NewInvalidValueObjectException creates an invalid value object exception
func NewInvalidValueObjectException(message, valueObjectType string, invalidValue interface{}) *InvalidValueObjectException {
	details := map[string]interface{}{
		"value_object_type": valueObjectType,
		"invalid_value":     fmt.Sprintf("%v", invalidValue),
	}

	exception := NewDomainException(11003001, message, details)

	return &InvalidValueObjectException{
		DomainException: exception,
		ValueObjectType: valueObjectType,
		InvalidValue:    invalidValue,
	}
}

// DomainErrorCodes contains domain error code constants
type DomainErrorCodes struct{}

var DomainErrors = DomainErrorCodes{}

// Business Rules (11001XXX)
const (
	BusinessRuleViolation   = 11001001
	UnderageUser            = 11001002
	VehicleTooOld           = 11001003
	InvalidVehicleCapacity  = 11001004
	MaxTripDurationExceeded = 11001005
)

// Entity State (11002XXX)
const (
	InvalidEntityState      = 11002001
	UserInactive            = 11002002
	DriverNotVerified       = 11002003
	TripAlreadyCompleted    = 11002004
	BookingAlreadyCancelled = 11002005
	LicenseExpired          = 11002006
)

// Value Objects (11003XXX)
const (
	InvalidValueObject     = 11003001
	InvalidCoordinates     = 11003002
	InvalidEmailFormat     = 11003003
	InvalidPhoneFormat     = 11003004
	InvalidRatingValue     = 11003005
	CoordinatesOutOfBounds = 11003006
)

// Aggregates (11004XXX)
const (
	AggregateConsistencyViolation = 11004001
	InvalidAggregateState         = 11004002
)

// Domain Services (11005XXX)
const (
	DomainServiceError       = 11005001
	DistanceCalculationError = 11005002
	PriceCalculationError    = 11005003
)

File: backbone-python/backbone/domain/exceptions/domain_exceptions.py
"""
Domain layer exceptions - Códigos 11XXXXXX
"""
from typing import Dict, Any, Optional
from .base_kernel_exception import BaseKernelException


class DomainException(BaseKernelException):
    """
    Excepción base para la capa de dominio.
    
    Códigos: 11XXXXXX
    HTTP por defecto: 400 (Bad Request)
    
    Uso: Violaciones de reglas de negocio, estados inválidos de entidades
    """
    
    def __init__(
        self, 
        code: int, 
        message: str, 
        http_code: int = 400,
        **kwargs
    ):
        # Validar que el código pertenezca a la capa de dominio
        if not (11000000 <= code <= 11999999):
            raise ValueError(f"Domain exception code must be in range 11000000-11999999, got: {code}")
        
        super().__init__(code, message, http_code, **kwargs)


class BusinessRuleViolationException(DomainException):
    """
    Excepción para violaciones de reglas de negocio.
    
    Ejemplos:
    - Usuario menor de edad
    - Vehículo demasiado antiguo  
    - Rating fuera de rango
    """
    
    def __init__(
        self,
        message: str,
        rule_name: str,
        code: int = 11001001,
        **kwargs
    ):
        details = kwargs.get('details', {})
        details.update({"rule_violated": rule_name})
        kwargs['details'] = details
        
        super().__init__(code, message, **kwargs)


class InvalidEntityStateException(DomainException):
    """
    Excepción para estados inválidos de entidades.
    
    Ejemplos:
    - Usuario inactivo intentando hacer una acción
    - Conductor no verificado
    - Viaje ya completado
    """
    
    def __init__(
        self,
        message: str,
        entity_type: str,
        entity_id: str,
        current_state: str,
        required_state: str,
        code: int = 11002001,
        **kwargs
    ):
        details = kwargs.get('details', {})
        details.update({
            "entity_type": entity_type,
            "entity_id": entity_id,
            "current_state": current_state,
            "required_state": required_state
        })
        kwargs['details'] = details
        
        super().__init__(code, message, http_code=412, **kwargs)


class InvalidValueObjectException(DomainException):
    """
    Excepción para value objects inválidos.
    
    Ejemplos:
    - Coordenadas fuera de rango
    - Email con formato inválido
    - Rating con valor inválido
    """
    
    def __init__(
        self,
        message: str,
        value_object_type: str,
        invalid_value: Any,
        code: int = 11003001,
        **kwargs
    ):
        details = kwargs.get('details', {})
        details.update({
            "value_object_type": value_object_type,
            "invalid_value": str(invalid_value)
        })
        kwargs['details'] = details
        
        super().__init__(code, message, **kwargs)


# Catálogo de códigos de dominio
class DomainErrorCodes:
    """Códigos de error específicos del dominio 11XXXXXX"""
    
    # Business Rules (11001XXX)
    BUSINESS_RULE_VIOLATION = 11001001
    UNDERAGE_USER = 11001002
    VEHICLE_TOO_OLD = 11001003
    INVALID_VEHICLE_CAPACITY = 11001004
    MAX_TRIP_DURATION_EXCEEDED = 11001005
    
    # Entity State (11002XXX)  
    INVALID_ENTITY_STATE = 11002001
    USER_INACTIVE = 11002002
    DRIVER_NOT_VERIFIED = 11002003
    TRIP_ALREADY_COMPLETED = 11002004
    BOOKING_ALREADY_CANCELLED = 11002005
    LICENSE_EXPIRED = 11002006
    
    # Value Objects (11003XXX)
    INVALID_VALUE_OBJECT = 11003001
    INVALID_COORDINATES = 11003002
    INVALID_EMAIL_FORMAT = 11003003
    INVALID_PHONE_FORMAT = 11003004
    INVALID_RATING_VALUE = 11003005
    COORDINATES_OUT_OF_BOUNDS = 11003006
    
    # Aggregates (11004XXX)
    AGGREGATE_CONSISTENCY_VIOLATION = 11004001
    INVALID_AGGREGATE_STATE = 11004002
    
    # Domain Services (11005XXX)
    DOMAIN_SERVICE_ERROR = 11005001
    DISTANCE_CALCULATION_ERROR = 11005002
    PRICE_CALCULATION_ERROR = 11005003